File: Linked List/Doubly_Linked_list.c
//Doubly Linked List
#include<stdio.h>
#include<stdlib.h>
typedef struct node {
    struct node *prev;
    int info;
    struct node *next;
}Node;


//Printing the list 
void printList(Node* start){
    Node *p;
    p = start;
    if(start == NULL){
        printf("List is empty\n");
        return;
    }
    while(p!=NULL){
        printf("%d ",p->info);
        p = p->next;
    }
    printf("\n");
    return;
}



//Insert in the empty list 

Node *add_to_empty(Node* start,int data){
    Node * tmp = (Node*)malloc(sizeof(Node));
    tmp->info = data;
    tmp->next = NULL;
    tmp->prev = NULL;
    start = tmp;
    return start;
}

//Insert at beg of the list
Node * add_at_beg(Node *start,int data){
    Node *tmp = (Node*)malloc(sizeof(Node));
    tmp->info = data;
    tmp->next = start;
    tmp->prev = NULL;
    start->prev = tmp;
    start = tmp;
    return start;
}


//Insert at end

Node* add_at_end(Node* start,int data){
    Node* tmp,*p;
    tmp =(Node*)malloc(sizeof(Node));
    tmp->info = data;
    p = start;
    while(p->next != NULL){
        p = p->next;
    }
    tmp->next = p->next;
    tmp->prev = p;
    p->next = tmp;
    return start;
}



int main(){
    Node *start = NULL;
    start = add_to_empty(start,1);
    start = add_at_end(start,2);
    start = add_at_end(start,3);
    printList(start);
    return 0;
}

File: Linked List/Java/DoublyLinkedList.java
class Node{
    int data;
    Node prev;
    Node next;
    public Node(int data){
        this.data = data;
    }
}
class LinkedList{
    Node head;
    Node tail;
    public void insert(int data){
        Node tmp = new Node(data);
        Node p = head;
        if(head == null){
            head = tmp;
            tmp.next = null;
            tmp.prev = null;
            tail = tmp;
            return;
        }
        while(p.next != null){
            p = p.next;
        }
        tail = tmp;
        tmp.next = null;
        tmp.prev = p;
        p.next = tmp;
        return;
    }
    
    public void insertStart(int data){
        Node tmp = new Node(data);
        tmp.next = head;
        tmp.prev = null;
        head.prev = tmp;
        head = tmp;
        return;
    }
    public void print(){
        Node p = head;
        if(head == null){
            System.out.println("List is empty");
            return;
        }
        p = head;
        while(p!= null){
            System.out.println(p.data+" ");
            p = p.next;
        }
    }
    public void printReverse(){
        Node p = tail;
        while(p != null){
            System.out.println(p.data);
            p = p.prev;
        }
    }
    
}
public class Main{
    public static void main(String args[]){
        LinkedList l = new LinkedList();
        l.insert(1);
        l.insert(2);
        l.insert(3);
        l.insert(4);
    }
}
